sdks/java/io/hadoop-input-format/src/main/java/org/apache/beam/sdk/io/hadoop/inputformat/HadoopInputFormatIO.java

```diff
@@ -0,0 +1,968 @@
+        this.valueTranslationFunction = valueTranslationFunction;
+        this.inputFormatObj = inputFormatObj;
+        this.split = split;
+      }
+
+      @Override
+      public HadoopInputFormatBoundedSource<K, V> getCurrentSource() {
+        return source;
+      }
+
+      private RecordReader<T, T> currentReader;
+      private KV<K, V> currentRecord;
+
+      @Override
+      public boolean start() throws IOException {
+        try {
+          TaskAttemptContextImpl attemptContext = new TaskAttemptContextImpl(
+              source.getConfiguration().getHadoopConfiguration(), new TaskAttemptID());
+          currentReader =
+              (RecordReader<T, T>) inputFormatObj.createRecordReader(split, attemptContext);
+            currentReader.initialize(split, attemptContext);
+            if (currentReader.nextKeyValue()) {
+              currentRecord = nextPair();
+              return true;
+            }
+        } catch (InterruptedException e) {
+          throw new IOException("Unable to read data : ", e);
+        }
+        currentReader = null;
+        currentRecord = null;
+        doneReading = true;
+        return false;
+      }
+
+      @Override
+      public boolean advance() throws IOException {
+        try {
+          synchronized (currentReader) {
+            if (currentReader != null && currentReader.nextKeyValue()) {
+              currentRecord = nextPair();
+              return true;
+            }
+          }
+          currentRecord = null;
+          doneReading = true;
+        } catch (InterruptedException e) {
+          throw new IOException("Unable to read data : ", e);
+        }
+        return false;
+      }
+
+      /**
+       * Returns the pair of key and value of the next record in sequence.
+       */
+      public KV<K, V> nextPair() throws IOException, InterruptedException {
+        // Transform key if required.
+        K key =
+            transformKeyOrValue(currentReader.getCurrentKey(), keyTranslationFunction, keyCoder);
+        // Transform value if required.
+        V value = transformKeyOrValue(currentReader.getCurrentValue(), valueTranslationFunction,
+            valueCoder);
+        recordsReturned++;
+        return KV.of(key, value);
+      }
+
+      /**
+       * Returns the serialized output by cloning. Given input is transformed with the provided
+       * SimpleFunction.
+       */
+      private <T1 extends Object> T1 transformKeyOrValue(T input,
+          @Nullable SimpleFunction<T, T1> simpleFunction, Coder<T1> coder)
+          throws IOException, InterruptedException {
+        T1 output;
+        if (null != simpleFunction) {
+          output = simpleFunction.apply(input);
+        } else {
+          output = (T1) input;
+        }
+        return clone(output, coder);
+      }
+
+      /**
+       * Cloning/Serialization is required to take care of Hadoop's mutable objects if returned by
+       * RecordReader as Beam needs immutable objects.
```

Comment:
1) there's a key piece of this explanation that's missing: many objects used by beam are mutable, but the Hadoop InputFormats tend to re-use the same object when returning them. 
2) this comment makes more sense in the transformKeyOrValue method - when you're in a function called clone, of course you are cloning. The question a dev looking at this code would have is why is it being called in transformKeyOrValue
3) This method is called clone, but it does not always clone. It should be called something that makes it clear to a call. Eg. cloneIfPossiblyMutable()